### PR TITLE
Make restoreGame async

### PR DESCRIPTION
**Pull request by kberg**

### CODE REVIEW

File: src/database/GameLoader.ts [hunk 1]
```diff
@@ -66,8 +66,22 @@ export class GameLoader implements IGameLoader {
     });
   }
 
+  public async getByParticipantIdAsync(id: PlayerId | SpectatorId): Promise<Game> {
+    const d = await this.idsContainer.getGames();
+    const gameId = d.participantIds.get(id);
+    if (gameId === undefined) {
+      return Promise.reject(new Error('Not found'));
```
Comment (bafolts) [inline src/database/GameLoader.ts:73]:
```suggestion
      throw new Error('Not found');
```

In an `async` function can `throw`. No need to return a `Promise`.

File: src/database/GameLoader.ts [hunk 2]
```diff
@@ -131,6 +132,38 @@ export class GameLoader implements IGameLoader {
     });
   }
 
+  private async loadGameAsync(gameId: GameId, bypassCache: boolean): Promise<Game> {
+    const d = await this.idsContainer.getGames();
+    const game = d.games.get(gameId);
+    if (bypassCache === false && game !== undefined) {
+      return game;
+    } else if (game === undefined) {
+      return Promise.reject(new Error(`GameLoader:game id not found ${gameId}`));
```
Comment (bafolts) [inline src/database/GameLoader.ts:141]:
```suggestion
      throw new Error(`GameLoader:game id not found ${gameId}`);
```

File: src/database/GameLoader.ts [hunk 3]
```diff
@@ -146,4 +179,17 @@ export class GameLoader implements IGameLoader {
       this.loadGame(gameId, false, cb);
     });
   }
+
+  private async loadParticipantAsync(id: PlayerId | SpectatorId): Promise<Game> {
+    const d = await this.idsContainer.getGames();
+    const gameId = d.participantIds.get(id);
+    if (gameId === undefined) {
+      return Promise.reject(new Error(`GameLoader:id not found ${id}`));
```
Comment (bafolts) [inline src/database/GameLoader.ts:187]:
```suggestion
      throw new Error(`GameLoader:id not found ${id}`);
```

File: src/database/PostgreSQL.ts [hunk 4]
```diff
@@ -213,30 +213,22 @@ export class PostgreSQL implements IDatabase {
...
-    });
+    const res = await this.client.query('SELECT game game FROM games WHERE game_id = $1 AND save_id = $2 ORDER BY save_id DESC LIMIT 1', [game_id, save_id]);
+    if (res.rows.length === 0) {
+      throw new Error(`Game ${game_id} not found`);
+    }
+    try {
+      // Transform string to json
+      const json = JSON.parse(res.rows[0].game);
+      logForUndo(json.id, 'restored to', json.lastSaveId, 'from', save_id);
+      return Promise.resolve(json);
```
Comment (bafolts) [inline src/database/PostgreSQL.ts:227]:
```suggestion
      return json;
```

File: src/routes/PlayerInput.ts [hunk 5]
```diff
@@ -1,45 +1,39 @@
...
   }
 
-  public override post(req: http.IncomingMessage, res: http.ServerResponse, ctx: IContext): void {
+  public override async post(req: http.IncomingMessage, res: http.ServerResponse, ctx: IContext): Promise<void> {
     const playerId = ctx.url.searchParams.get('id');
 
     if (playerId === null) {
       ctx.route.badRequest(req, res, 'must provide player id');
-      return;
+      return Promise.resolve();
```
Comment (bafolts) [inline src/routes/PlayerInput.ts:21]:
```suggestion
      return;
```

File: src/routes/PlayerInput.ts [hunk 6]
```diff
@@ -1,45 +1,39 @@
...
+    const game = await ctx.gameLoader.getByParticipantIdAsync(playerId);
+    let player: Player | undefined;
+    try {
+      player = game.getPlayerById(playerId);
+    } catch (err) {
+      console.warn(`unable to find player ${playerId}`, err);
+    }
+    if (player === undefined) {
+      ctx.route.notFound(req, res);
+      return Promise.resolve();
```
Comment (bafolts) [inline src/routes/PlayerInput.ts:34]:
```suggestion
      return;
```

**Review by bafolts — APPROVED** [review]

**Comment by kberg** [comment]
Thanks for all the suggestions!

**Pull request merged**